aether-ui/aether/ui/api/utils.py

```diff
@@ -171,212 +169,147 @@ def kernel_data_request(url='', method='get', data=None):
+            outcome['exists'].append({contract.name: 'Pipeline (mapping) with name {} exists on kernel.'.format(
+                                    contract.name)})
+            outcome['ids']['mapping'][contract.name] = get_by_name[0]['id']
+
+    if pipeline.mappingset:
+        try:
+            mappingset = kernel_data_request(f'mappingsets/{pipeline.mappingset}/')
+            if json.dumps(mappingset['input'], sort_keys=True) != json.dumps(pipeline.schema, sort_keys=True):
+                outcome['exists'].append({pipeline.name: 'Input data will be changed'})
+        except Exception:
+            pass
     return outcome
 
 
-def publish_pipeline(pipeline, projectname, overwrite=False):
+def publish_pipeline(pipeline, project_name, contract, objects_to_overwrite={}):
     '''
-    Transform pipeline to kernel data and publish
+    Transform pipeline and contract to kernel artefacts and publish
     '''
-    outcome = {
-        'successful': [],
-        'error': [],
-        'exists': []
-    }
-    pipeline.kernel_refs = pipeline.kernel_refs if pipeline.kernel_refs else {}
-    project_data = {
-        'revision': str(uuid.uuid4()),
-        'name': projectname,
-        'salad_schema': '[]',
-        'jsonld_context': '[]',
-        'rdf_definition': '[]'
-    }
+    project_id = str(uuid.uuid4())
+    mappingsets = []
+    mappings = []
+    schemas = []
+    outcome = {}
     try:
-        create_new_kernel_object('project', pipeline, project_data)
-        outcome['successful'].append('{} project created'.format(projectname))
-    except Exception:
-        if overwrite and 'project' in pipeline.kernel_refs:
-            update_kernel_object('project', pipeline.kernel_refs['project'], project_data)
-            outcome['successful'].append('{} project updated'.format(projectname))
-        else:
-            get_by_name = kernel_data_request(f'projects/?name={projectname}')['results'][0]
-            pipeline.kernel_refs['project'] = get_by_name['id']
-            outcome['successful'].append('Existing {} project used'.format(projectname))
-
-    for entity_type in pipeline.entity_types:
-        schema_name = entity_type['name']
-        schema_data = {
-            'revision': str(uuid.uuid4()),
-            'name': schema_name,
-            'type': entity_type['type'],
-            'definition': entity_type
-        }
+        projects = kernel_data_request(f'projects/?name={project_name}')['results']
+        aux_project = projects[0] if len(projects) else {}
+        project_id = aux_project.get('id', str(uuid.uuid4()))
+        project_name = aux_project.get('name', 'Aux')
+    except Exception as e:
+        pass
+    if contract.kernel_refs and 'project' in contract.kernel_refs:
+        try:
+            project = kernel_data_request(f'projects/{contract.kernel_refs["project"]}/')
+            project_id = project.get('id')
+            project_name = project.get('name')
+        except Exception:
+            pass
```

Comment:
I think that the project id should be extracted from the pipeline mapping set and not kept within the contract...

```python
    # check linked Project I (using pipeline mapping set)
    project_id = None
    mappingset_id = None
    
    if pipeline.mappingset:
        mappingset_id = pipeline.mappingset
        try:
            kernel_mappingset = kernel_data_request(f'mappingsets/{pipeline.mappingset}/')
            project_id = mappingset.get('project')
            # take from kernel_mappingset the data to be push back to kernel
        except Exception:
            # take from pipeline the data to be pushed to kernel
            pass
    else:
        mappingset_id = str(uuid.uuid4())
        # take from pipeline the data to be pushed to kernel

    # check linked Project II (using given project name)
    if not project_id and project_name:
        try:
            projects = kernel_data_request(f'projects/?name={project_name}')['results']
            aux_project = projects[0] if len(projects) else {}
            project_id = aux_project.get('id')
        except Exception as e:
            pass

    # check linked Project III (if everything else failed...)
    if not project_id:
        project_id = str(uuid.uuid4())
```

and continue with contract...

Reply:
The algorithm here is to retrieve the linked project using `pipeline.mappingset` or `contract.kernel_refs['project']` if previously publish, or create and link an `Aux` project to  all new pipelines coming from the ui